pyro/infer/mcmc/mcmc.py

```diff
@@ -137,6 +135,17 @@ def terminate(self):
                 w.terminate()
 
     def _traces(self, *args, **kwargs):
+        chain_indices = [[] for _ in range(self.num_chains)]
+        try:
```

Comment:
I think we should modify the TracePosterior class itself to store chain information, and plug in diagnostics there. I am fine with merging this if we need it earlier, but we should refactor this soon. I can take a look at it this week.